feat: use  `manifest-beta` for BRAT pre-releases

johannrichard opened the pull request:
Thanks for a great plugin! I implemented a small change that will create a `manifest-beta.json` in case a pre-release is detected (e.g. when committing to a `beta` or `next` branch). This will then create a `manifest-beta.json` for that pre-release and commit it to the branch, but will leave the `manifest.json` untouched.

Because of the way BRAT (currently) works, you either have to set your pre-release branch as the default branch and update it's `manifest.json` manually whenever a new version is published, or copy the `manifest-beta.json` from your pre-release branch to the default branch. 

I'm not sure which one is the best way, or if there are better ways to solve this (including updating BRAT such that a branch could be specified instead of `manifest-beta.json` or a fixed version number).

johannrichard committed:
feat: ✨ use  `manifest-beta` for BRAT pre-releases

- use the `manifest-beta.json` for pre-releases
- leave `manifest.json` untouched for pre-releases

johannrichard committed:
feat: ✨ create `manifest-beta.json` from current `manifest.json` in case of pre-release

johannrichard committed:
docs: 📝 update README to include beta release support and manifest-beta.json details

johannrichard commented:
I submitted https://github.com/TfTHacker/obsidian42-brat/pull/93 to `obsidian42-brat`. In case it is integrated, the code required to create, update and commit the `manifest-beta.json` would not be needed anymore, and the problem outlined above would vanish. Happy to keep that PR open as aa draft as long as this question is not solved.

brianrodri commented:
This looks great, thanks for the contribution! I'd be happy to merge this in but please add a `TODO(TfTHacker/obsidian42-brat#93)` comment somewhere so that I can remember to remove it once and if your PR goes through.

I'm not a big user of BRAT, so let me know what'd improve your experience the most. Thanks again and happy new years! 🎉

brianrodri committed:
Merge branch 'main' into brat-prerelease

brianrodri committed:
Merge branch 'main' into brat-prerelease

brianrodri commented:
Heads up I merged in some changes to your branch so that it's passing my checks!

johannrichard committed:
chore: add TODO vis-à-vis BRAT

TODO(TfTHacker/obsidian42-brat#93): Make sure to remove code if BRAT integrates changes to use GitHub pre-releases

johannrichard commented:
@brianrodri I have continued playing around with the plugin. It's really great! I have written up a small Gist with a sample `.releaserc.yaml` which contains a boilerplate configuration that should work with minimal adjustments for almost every Obsidian plugin. This works without polluting the `package.json` (which is especially useful when tinkering with (development) forks, for which one can now release test versions with a breeze).

I also wrote up a small [README.md](https://gist.github.com/johannrichard/b75ec94659d2d3db4b7b3bcf7fe2a8c1#file-readme-md) on [Using `semantic-release` for your Obsidian plugin and development forks alike](https://gist.github.com/johannrichard/b75ec94659d2d3db4b7b3bcf7fe2a8c1#file-readme-md) that briefly explains how to use your plugin in different contexts.

brianrodri commented on src/prepare.js:
```diff
@@ -1,16 +1,31 @@
             case "versions.json":
                 json[version] = minAppVersion;
                 break;
+            case "manifest.json":
+                // Don't touch manifest.json in case of pre-release
+                if (!isPrerelease(context.branch)) {
+                    json.version = version;
+                }
+                break;
```
This script assumes that the JSON version has been updated, so I think it makes sense to add a case for `manifest-beta.json` so that the version bump is at least captured there.

What do you think?

brianrodri commented on src/prepare.js:
```diff
@@ -1,16 +1,31 @@
+    // TODO(TfTHacker/obsidian42-brat#93): Remove if BRAT updates to GitHub release based workings
+    // Pre-release, add `manifest-beta.json` key to filemap
+    if (isPrerelease(context.branch)) {
+        fileMap.set("manifest-beta.json", {
+            ...fileMap.get("manifest.json"),
+        });
```
I'm not too familiar with BRAT, but I have seen `manifest-beta.json` files laying around in plugin repos.

Would it make sense, from your desired workflow's point-of-view, to load the pre-existing `manifest-beta.json` if it exists? Or is it always OK to clobber the contents w/ whatever the `manifest.json` contains?

johannrichard replied:
To my understanding, the `manifest-beta.json` is only used to indicate the beta version to BRAT. However, for the sake of consistency, I think it is best to use `manifest.json` as the source from which we create the `manifest-beta.json`.

brianrodri commented on src/is-prerelease.js:
```diff
@@ -0,0 +1,8 @@
+    if (prerelease === false) {
+        return false;
+    }
+    return (
```
This seems a little complicated, could you please split these into a chain of if-else with brief comments describing how a user might "land" in each branch?

Also, please add some unit tests to demonstrate the same ideas (and so codecov stays at 100% 😊)

brianrodri commented on README.md:
```diff
@@ -36,4 +36,21 @@ When `semantic-release` runs, this plugin will update the following files accord
+
+## Beta Release Support with BRAT
+
+When running `semantic-release` on a pre-release branch (e.g. `beta`), the plugin updates only `manifest-beta.json` and `versions.json` with the beta-version number, preserving `manifest.json`. This enables [BRAT](https://github.com/TfTHacker/obsidian42-brat) compatibility.
```
I don't see `beta` hard-coded anywhere, does that mean _any_ branch name would work? If so, please update to communicate that.

brianrodri replied:
```suggestion
When running `semantic-release` on a pre-release branch (e.g. `beta`), the plugin updates only `manifest-beta.json` and `versions.json` with the beta-version number, preserving `manifest.json`. This enables [BRAT](https://github.com/TfTHacker/obsidian42-brat) compatibility. Note that _any_ branch can act as a pre-release branch, just make sure it's not configured as a "release" branch in the GitHub workflows.

brianrodri reviewed (changes requested):
Got it, thanks @johannrichard! I've gone ahead and given the code changes a genuine review, PTAL!

I'll also play around with the .yml snippet on a private repository to quickly confirm, then I'll open a PR myself with the changes. Would you mind reviewing those changes when I get around to it?

Finally, thanks again!!

johannrichard committed:
docs: update README.md

brianrodri commented on src/prepare.js:
```diff
@@ -1,16 +1,31 @@
             case "versions.json":
                 json[version] = minAppVersion;
                 break;
+            case "manifest.json":
+                // Don't touch manifest.json in case of pre-release
+                if (!isPrerelease(context.branch)) {
+                    json.version = version;
+                }
+                break;
```
Shouldn't we still bump `manifest-beta.json`'s version so that BRAT detects that a new version is available? Otherwise it seems like it'll never "see" the new versions since it's always copying from `manifest.json`, which isn't being changed.

johannrichard replied:
Unless I'm completely missing something, this should already be happening: if we work on a pre-release, we push `manifest-beta.json` to the map as a copy of `manifest.json` before we run the code to update the version in `manifest-beta.json`. If we're on a normal release, we won't touch `manifest-beta.json` but update `manifest.json` instead.

brianrodri replied:
Ah right, my own code confused me 😎 I forgot that the default case statement will handle it. Please disregard!

johannrichard committed:
docs: update README.md

Add some more detail on pre-release branches.

johannrichard committed:
docs: fix README.md

The plugin must be specified as `semantic-release-obsidian-plugin` in the `plugins` list. Otherwise, `semantic-release` will fail.

johannrichard commented:
PS: I saw there's [a longer thread in the Obsidian Forum on GitHub actions](https://forum.obsidian.md/t/using-github-actions-to-release-plugins/7877/16). None of them mention your `semantic-release` plugin, and all of them involve quite complex setups with additional scripts. I was thinking a) of replying by adding a pointer to your plugin (b/c it's really great and works well out of the box 🥳) and b) of shamelessly adding a link to my small write-up (relevant parts of which we can also integrate into your plugin's doc if you prefer). 

Please let me know what you think of a) and b).

brianrodri commented:
Haha b) sounds good to me!

They'll need a link to this repo to file feature requests anyway 😎

brianrodri committed:
Merge branch 'main' into brat-prerelease

johannrichard commented:
BRAT now supports GitHub releases and integrated the changes proposed earlier so this PR is not anymore needed. Closing it for now.
